Batching for AvaloniaObject property values.

grokys opened the pull request:
## What does the pull request do?

Cuts down the amount of notifications raised when controls' styling is attached/detached:

- Adds batching to `AvaloniaObject` by calling `BeginBatchUpdate`/`EndBatchUpdate`. During batch update, no property change notifications are sent and bindings are not subscribed
- Don't subscribe to `PropertySetterBindingInstance`'s inner observable if the containing style instance is not active: this reduces unnecessary resource lookups
- Lazily evaluate `<Template>`s in setters

Part of fixing #5027.

Not likely to be ready for 0.10.0 - will need a fair bit of testing.

## Benchmarks

before

```
|       Method |     Mean |     Error |    StdDev |   Gen 0 |   Gen 1 | Gen 2 | Allocated |
|------------- |---------:|----------:|----------:|--------:|--------:|------:|----------:|
| RepeatButton | 1.288 ms | 0.0209 ms | 0.0185 ms | 31.2500 | 13.6719 |     - |  292.1 KB |
```

after

```
|       Method |     Mean |   Error |   StdDev |   Median |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|--------:|---------:|---------:|--------:|------:|------:|----------:|
| RepeatButton | 391.2 us | 7.77 us | 18.77 us | 382.6 us | 15.1367 |     - |     - |  62.52 KB |
```


## Checklist

- [x] Added unit tests (if possible)?

grokys committed:
Don't subscribe to inner observable if not active.

When subscribing to a `PropertySetterBindingInstance`, if the owner style is not active then there's no need to subscribe to the inner observable as this can cause resource lookups etc. Part of fixing #5027.

grokys committed:
Added failing tests for #5027.

We shouldn't subscribe to bindings until needed.

grokys committed:
Lazily evaluate <Template>s in setters.

Part of fixing #5027.

grokys committed:
Added benchmark for fluent RepeatButton.

As that's where #5027 was showing up most.

grokys committed:
More tests for #5027.

grokys committed:
Implement batching for AvaloniaObject property values.

Cuts down the amount of notifications raised when controls' stying is attached/detached.

Part of fixing #5027.

grokys commented on src/Avalonia.Base/PropertyStore/BindingEntry.cs:
```diff
@@ -22,6 +23,8 @@ internal class BindingEntry<T> : IBindingEntry, IPriorityValueEntry<T>, IObserve
         private readonly IAvaloniaObject _owner;
         private IValueSink _sink;
         private IDisposable? _subscription;
+        private bool _isSubscribed;
```
Probably worth packing these into an flags enum to save 4 bytes per value stored.

MarchingCube replied:
If only life would be that simple 😄 Depending on the type of the entry alignment of this class will change, for example [Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoa0gAgCEBLPAExYHMBRPDKATwA8AFQB8dAO4ALGLDrC6IOgGc+OMBmoBvanT10ADlCYA3AIYYYdWGdYQ8AG350A8sABWMDXQD6ECXiyANy6+kamFlZunt4+yiwA1iFU+obG5pZ0EB5eGL7KOMDKYMYGGEz2yanhGVbAEBAOvkzKAMqFxcbAMKxVYemRdPWNvsAWYFIAqgaskX16NYPCAPy+5g44MPN0oXpEAMyuOd6tEDhQYFZaHDAYQQC+1I80VKq43gBqXvvau3QHdAAImdgA4rAANbYA4GFMF0ACaUMOMNBVgAWslnrRDvQAMJ0HQpfSkEgEv6pcn6ACSeGUBlyADoABIwMwGAAUgQkjBY7Dw3F4AkELAwonZAEpxdtUjS6YyWWzOTBucw2JweHwhCiwWLJdLqbT6RpmayOVyeWr+RqhV8wPtdVK/s97kA) reducing size will do nothing since there is no free space and it will get aligned.

grokys commented on src/Avalonia.Base/PropertyStore/PriorityValue.cs:
```diff
@@ -18,14 +18,16 @@ namespace Avalonia.PropertyStore
         private Optional<T> _localValue;
         private Optional<T> _value;
+        private bool _isCalculatingValue;
+        private bool _batchUpdate;
```
Probably worth packing these into an flags enum to save 4 bytes per value stored.

grokys committed:
Removed old code.

grokys committed:
Remove unused field.

grokys committed:
Fix LocalValue bindings during batch update.

LocalValue bindings are special case due to their interaction with the local value set by a non-binding.

grokys committed:
Handle batching binding completion.

grokys committed:
Handle nested batch updates.

Commit:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

grokys committed:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

grokys committed:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

MarchingCube committed:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

Commit:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

Commit:
fix build.

MarchingCube commented on src/Avalonia.Styling/Styling/Setter.cs:
```diff
@@ -97,6 +89,13 @@ public ISetterInstance Instance(IStyleable target)
+                data.result = new PropertySetterLazyInstance<T>(
+                    data.target,
+                    property,
+                    () => (T)template.Build());
```
perf: eliminate closure and delegate allocation.

MarchingCube commented on src/Avalonia.Styling/Styling/Setter.cs:
```diff
@@ -117,6 +116,13 @@ public ISetterInstance Instance(IStyleable target)
+                data.result = new PropertySetterLazyInstance<T>(
+                    data.target,
+                    property,
+                    () => (T)template.Build());
```
perf: eliminate closure and delegate allocation.

MarchingCube commented on tests/Avalonia.Benchmarks/Themes/FluentBenchmark.cs:
```diff
@@ -0,0 +1,74 @@
+            AssetLoader.RegisterResUriParsers();
+            return new Styles
+            {
+                new StyleInclude(new Uri("avares://Avalonia.Benchmarks"))
```
I guess this can be replaced now by `new FluentTheme()`?

MarchingCube reviewed (approved):
I've been running this branch in my app for a few days already. I've definitely noticed that resource lookup count is more manageable and random templates are not being instantiated.

Left two perf comments but mostly for keeping track of it, I can fix it later (such setters are quite rare anyway).

MarchingCube committed:
Merge branch 'master' into fixes/5027-avaloniaobject-batching

grokys committed:
Merge branch 'master' into fixes/5027-avaloniaobject-batching
